lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -1595,6 +1649,169 @@ class BuildBBRegs : public FuncOpPartialLoweringPattern {
+        // until after BuildOpGroups, and RAUW inside BuildPipelineGroups, once
+        // the pipeline register created here has been assigned.
+      }
+    });
+    return success();
+  }
+};
+
+/// Builds groups for assigning registers for pipeline stages.
+class BuildPipelineGroups : public FuncOpPartialLoweringPattern {
+  using FuncOpPartialLoweringPattern::FuncOpPartialLoweringPattern;
+
+  LogicalResult
+  PartiallyLowerFuncToComp(mlir::FuncOp funcOp,
+                           PatternRewriter &rewriter) const override {
+    for (auto pipeline : funcOp.getOps<staticlogic::PipelineWhileOp>())
+      for (auto stage :
+           pipeline.getStagesBlock().getOps<staticlogic::PipelineStageOp>())
+        if (failed(buildStageGroups(stage, rewriter)))
+          return failure();
+
+    return success();
+  }
+
+  LogicalResult buildStageGroups(staticlogic::PipelineStageOp stage,
+                                 PatternRewriter &rewriter) const {
+    // Collect pipeline registers for stage.
+    auto pipelineRegisters = getComponentState().getPipelineRegs(stage);
+
+    // Collect group names for the prologue or epilogue.
+    SmallVector<StringAttr> prologueGroups;
+    SmallVector<StringAttr> epilogueGroups;
+    auto whileOp = stage->getParentOfType<staticlogic::PipelineWhileOp>();
+
+    // For each registered stage result value.
+    auto stageResults = stage.getBodyBlock().getTerminator()->getOperands();
+    for (size_t i = 0, e = stageResults.size(); i < e; ++i) {
+      auto value = stageResults[i];
+
+      // Get the pipeline register for that result.
+      auto pipelineRegister = pipelineRegisters[i];
+
+      // Get the evaluating group for that value.
+      auto evaluatingGroup = getComponentState().getEvaluatingGroup(value);
+
+      // Remember the final group for this stage result.
+      calyx::GroupOp group;
+
+      // Stitch the register in, depending on whether the group was
+      // combinational or sequential.
+      if (auto combGroup =
+              dyn_cast<calyx::CombGroupOp>(evaluatingGroup.getOperation())) {
+        // Create a sequential group and replace the comb group.
+        PatternRewriter::InsertionGuard g(rewriter);
+        rewriter.setInsertionPoint(combGroup);
+        group = rewriter.create<calyx::GroupOp>(combGroup.getLoc(),
+                                                combGroup.getName());
+        rewriter.cloneRegionBefore(combGroup.getBodyRegion(), group.getBody());
+        group.getBodyRegion().back().erase();
+        rewriter.eraseOp(combGroup);
+
+        // Stitch evaluating group to register.
+        buildAssignmentsForRegisterWrite(getComponentState(), rewriter, group,
+                                         pipelineRegister, value);
+
+        // Register all sources as having been made sequential.
+        for (auto assign : group.getOps<calyx::AssignOp>())
+          getComponentState().addSeqGroupValue(assign.src());
```

Comment:
This is one place that was a little hacky, and I'm not sure what to do about it. The simplest way I could think to implement the lowering here was to turn comb groups into seq groups. The only problem this caused was this breaks InlineCombGroups, which crashes when it encounters an op that was previously in a comb group but no longer is. I opted to add this extra bit of state to bail out in such cases in InlineCombGroups, but I'm not sure if there is a cleaner way to achieve this behavior.

Reply:
> which crashes when it encounters an op that was previously in a comb group but no longer is.

This sounds to me like an issue with some state not being updated when converting the group type?

Reply:
The problem is InlineCombGroups has a small list of know sequential ops, and if the assignment is not from one of them, it calls `state.getEvaluatingGroup<calyx::CombGroupOp>`, which fails. For example, an addition would not be in the list, we'd call getEvaluatingGroup, but the addition has been moved into a calyx::GroupOp. Is there some existing state I should be updating? My solution was to add this new bit of state.

Reply:
Ok, I think I see the better way to do this. I don't need any extra state, what was missing was calling `registerEvaluatingGroup` when a comb group is turned into a sequential one. With that, I can turn the assertion in InlineCombGroups into a condition, and all else is satisfied.

Reply:
Updated in 4217715